_source/_posts/2022-04-26-webauthn-java.md

```diff
@@ -0,0 +1,713 @@
+
+```xml
+<dependency>
+    <groupId>com.yubico</groupId>
+    <artifactId>webauthn-server-core</artifactId>
+    <version>1.12.1</version>
+</dependency>
+```
+
+## Data layer
+
+The application stores two data object classes: users and credentials. Users are people using the system, and they can have multiple credentials. Credentials contain the information needed to identify and verify a device implementing Client to Authenticator (CTAP2) protocol. CTAP2 is a spec that describes communication between a roaming authenticator and another client/platform at the application layer, as well as bindings to a variety of transport protocols that use different physical media.
+
+### User data
+
+There are two goals for the user data: track the existence and uniqueness of the user and their credentials, and enable the creation of two JavaScript objectsdictionaries for the browser to make WebAuthn API requests: [`PublicKeyCredentialCreationOptions`](https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer/create) for registration and [`PublicKeyCredentialRequestOptions`](https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredentialRequestOptions) for authentication.
+
+Using Spring JPA, the `AppUser` data class starts with traditional fields for organizing users in a system, an `id` for database lookup, and a unique `username` to enable user self-identification.
+
+There are also fields that the server needs to create the WebAuthn requests:
+
+```java
+import com.yubico.webauthn.data.ByteArray;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Lob;
+
+@Entity
+@Getter
+@NoArgsConstructor
+public class AppUser {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private Long id;
+
+    @Column(nullable = false, unique = true)
+    private String username;
+
+    @Column(nullable = false)
+    private String displayName;
+
+    @Lob
+    @Column(nullable = false, length = 64)
+    private ByteArray handle;
+}
+```
+
+The name field is set by the user. It is intended to be displayed by UI elements on the client as part of the registration and authentication process, but it's also entirely optional in the WebAuthn process. According to the [WebAuthn Specification](https://www.w3.org/TR/webauthn-1/#sctn-user-handle-privacy), a nickname string like `username` is not suitable for identification, so WebAuthn identity requests are made using a user handle, a byte sequence with a maximum length of 64.
+
+Byte data (like the user handle) is stored in the database as a Binary Large Object. To convert these data fields to the `ByteArray` object the application uses, the application implements a JPA `AttributeConverter` class with the `Converter` annotation.
```

Comment:
```suggestion
Byte data (like the user handle) is stored in the database as a binary large object (BLOB). To convert these data fields to the `ByteArray` object the application uses, the application implements a JPA `AttributeConverter` class with the `Converter` annotation.
```

Because https://en.wikipedia.org/wiki/Binary_large_object